include/tvm/tir/sparse.h

```diff
@@ -99,23 +99,48 @@ class DenseAxis : public Axis {
   TVM_DEFINE_OBJECT_REF_METHODS(DenseAxis, Axis, DenseAxisNode);
 };
 
+/*!
+ * \brief Sparse axis whose column indices is not consecutive.
+ */
+class SparseAxisNode : public AxisNode {
+ public:
+  static constexpr const char* _type_key = "tir.sparse.SparseAxis";
+  TVM_DECLARE_BASE_OBJECT_INFO(SparseAxisNode, AxisNode);
+};
+
+/*!
+ * \brief Managed reference to SparseAxisNode.
+ * \sa SparseAxisNode
+ */
+class SparseAxis : public Axis {
+ public:
+  TVM_DEFINE_OBJECT_REF_METHODS(SparseAxis, Axis, SparseAxisNode);
+};
+
 /*!
  * \brief Dense axis with fixed length per row.
  */
 class DenseFixedAxisNode : public DenseAxisNode {
  public:
+  Optional<SparseAxis> from_sparse;
+
   void VisitAttrs(AttrVisitor* v) {
     v->Visit("name", &name);
     v->Visit("length", &length);
+    v->Visit("from_sparse", &from_sparse);
   }
 
-  bool SEqualReduce(const DenseAxisNode* other, SEqualReducer equal) const {
-    return equal(name, other->name) && equal(length, other->length);
+  bool SEqualReduce(const DenseFixedAxisNode* other, SEqualReducer equal) const {
+    equal->MarkGraphNode();
```

Comment:
Could you briefly explain how does the `MarkGraphNode` work?

Reply:
When `MarkGraphNode` is on, the pair of Objects to be compared will be marked as `graph_equal = true`:
https://github.com/yzh119/tvm/blob/fb2c49fd9684c87d1b922e66784997cac170bb1a/src/node/structural_equal.cc#L92-L96

In this situation, when doing structural equality check, the checker will put the equivalent pair into two internal map containers:
https://github.com/yzh119/tvm/blob/fb2c49fd9684c87d1b922e66784997cac170bb1a/src/node/structural_equal.cc#L150-L153

And the mappings will provide equivalence information when we do structural equality check over two `Map`, even when their items have different relative orders:
https://github.com/yzh119/tvm/blob/fb2c49fd9684c87d1b922e66784997cac170bb1a/src/node/structural_hash.cc#L463-L476

_The method `MapLhsToRhs(...)` simply looks up the mappings:_
https://github.com/yzh119/tvm/blob/fb2c49fd9684c87d1b922e66784997cac170bb1a/src/node/structural_equal.cc#L98-L103

---

Finally, we can safely do structural equality check among mappings `sp_struct2param_map` without worrying about the relative orders of the mapping items 😉.

Reply:
That make sense to me, btw why don't we maintain a union-find set for equivalence check?

Reply:
In my understanding structural equality check only compares two objects once at a time. I think union-find set will be more helpful to represent equivalence in among many objects.

I'm not the designer of this check, though. But if I was asked to write such a check, I'd also like to keep the check as lightweight as possible (I mean, keep little information as possible).